src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs

```diff
@@ -464,9 +467,69 @@ public LineBatch GetBatch()
                     throw Contracts.ExceptDecode(batch.Exception, "Stream reading encountered exception");
                 }
 
+                private static class MultiLineReader
+                {
+                    // When reading lines that contain quoted fields, the quoted fields can contain
+                    // '\n' so we we'll need to read multiple lines (multilines) to get all the fields
+                    // of a given row.
+                    public static string ReadMultiLine(TextReader sr, StringBuilder sb, bool ignoreHashLine)
+                    {
+                        string line;
+                        line = sr.ReadLine();
+
+                        // if it was an empty line or if we've reached the end of file (i.e. line = null)
+                        if (string.IsNullOrEmpty(line))
+                            return line;
+
+                        // In ML.NET we filter out lines beginning with // and # at the beginning of the file
+                        // Or lines beginning with // elsewhere in the file.
+                        // Thus, we don't care to check if there's a quoted multiline when the line begins with
+                        // these chars.
+                        if (line[0] == '/' && line[1] == '/')
+                            return line;
+                        if (ignoreHashLine && line[0] == '#')
+                            return line;
```

Comment:
Should these check if you're at the top of the file? I think once you've hit the first row of data, no more ignoring of rows is allowed.

Reply:
Current behavior is that at the beginning of the file both // and # are ignored by LineReader [link](https://github.com/dotnet/machinelearning/blob/212081c62855a835e0612887f705a8f87a899624/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs#L494-L495), but after the first row of data is hit, LineReader only ignores // [link](https://github.com/dotnet/machinelearning/blob/212081c62855a835e0612887f705a8f87a899624/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs#L526-L534)

Because of this, I actually added the ignoreHashLine flag on this method, and use it on the code you've pointed to.

One of the tests I've added on this PR actually shows that "// lines" are ignored correctly throughout the file.

---
In reply to: [425748062](https://github.com/dotnet/machinelearning/pull/5125#discussion_r425748062) [](ancestors = 425748062)